### PR TITLE
fix: handle both 1D and 2D input shapes in XGBoost custom objectives

### LINKED ISSUE #865: feat(compat): add XGBoost 3.2 compatibility for custom objectives
## Background

XGBoost 3.2 (Feb 2026) introduced significant changes to multi-target tree support with vector-leaf and changes to how multi-output data flows through the sklearn interface's `_objective_decorator`. This broke OpenSTEF's custom objective functions.

As a **temporary fix**, XGBoost was pinned to `<3.2` in #864. This issue tracks the proper fix.

## Root Cause

The custom objectives (`arctan_loss_multi_objective`, `pinball_loss_multi_objective`) in `packages/openstef-models/src/openstef_models/utils/loss_functions.py` assume XGBoost passes a **1D flattened array** to the objective:

```python
n_items = len(y_true)
n_quantiles = len(quantiles)
n_rows = n_items // n_quantiles          # ← fragile assumption
y_pred = np.reshape(y_pred, (n_rows, n_quantiles))
y_true = np.reshape(y_true, (n_rows, n_quantiles))
```

Similarly in `mean_pinball_loss` (evaluation function):
```python
y_pred = np.reshape(y_pred, [-1, n_quantiles])
```

XGBoost 3.2 may pass **2D `(n_samples, n_quantiles)` arrays** directly to the custom objective when using `multi_strategy="one_output_per_tree"`, making the manual reshape unnecessary or wrong.

Also need to verify that `sample_weight_eval_set` still works correctly in 3.2.

## Scope

Files to investigate and update:
- `packages/openstef-models/src/openstef_models/utils/loss_functions.py` — objective functions (`arctan_loss_multi_objective`, `pinball_loss_multi_objective`, `xgb_prepare_target_for_objective`)
- `packages/openstef-beam/src/openstef_beam/metrics/metrics_probabilistic.py` — `mean_pinball_loss` (used as eval metric)
- `packages/openstef-models/src/openstef_models/models/forecasting/xgboost_forecaster.py` — `fit()` / `predict()` reshape logic

## Acceptance Criteria

- [ ] A minimal reproduction script that demonstrates the failure with XGBoost 3.2
- [ ] Objective functions handle both 1D (old) and 2D (new) input shapes, or exclusively use the 2D API
- [ ] All existing XGBoost forecaster tests pass with XGBoost 3.2 installed
- [ ] Version constraint in `packages/openstef-models/pyproject.toml` relaxed back to `>=3,<4` (or `>=3.2,<4` if minimum is bumped)
- [ ] `uv.lock` updated

## Notes

- XGBoost 3.2 changelog: https://xgboost.readthedocs.io/en/latest/changes/v3.2.0.html
- Advanced custom objectives guide: https://xgboost.readthedocs.io/en/latest/tutorials/advanced_custom_obj.html
- The `multi_strategy="one_output_per_tree"` param docs: https://xgboost.readthedocs.io/en/latest/tutorials/multioutput.html
- `_objective_decorator` in the sklearn interface wraps the custom fn and passes `dmatrix.get_label()` / weights to it — check what shape this returns in 3.2 for multi-output models

### PR DESCRIPTION
**Pull request by Soutehkeshan**
## Summary

Fixes compatibility with XGBoost 3.2, which changed the shape of arrays passed to custom objectives via the sklearn interface. With `multi_strategy="one_output_per_tree"`, `y_true` and `y_pred` are now passed as 2D `(n_samples, n_quantiles)` arrays instead of 1D flattened arrays of length `n_samples * n_quantiles`.

Closes #865 

## Root Cause

The old reshape logic in both objective functions assumed 1D input:

```python
n_items = len(y_true)
n_quantiles = len(quantiles)
n_rows = n_items // n_quantiles  # ← breaks with 2D input
y_pred = np.reshape(y_pred, (n_rows, n_quantiles))
y_true = np.reshape(y_true, (n_rows, n_quantiles))
sample_weight = np.reshape(sample_weight, (n_rows, -1)) if sample_weight is not None else None
```

With 2D input, `len(y_true)` returns `n_samples` (not `n_samples * n_quantiles`), so `n_rows = n_samples // n_quantiles` is wrong, causing the reshape to fail or produce incorrect results.

The below image shows running a minimal script using XGBoost 3.1:
Screenshot 2026-05-04 at 21 19 23

The below image shows running the same minimal script using XGBoost 3.2:
Screenshot 2026-05-04 at 21 22 12

## Fix

Replaced with shape-agnostic normalisation in `arctan_loss_multi_objective` and `pinball_loss_multi_objective`:

```python
n_quantiles = len(quantiles)
y_pred = np.reshape(y_pred, (-1, n_quantiles))
y_true = np.reshape(y_true, (-1, n_quantiles))
sample_weight = sample_weight[:, np.newaxis] if sample_weight is not None else None
```

`np.reshape(arr, (-1, n_quantiles))` is a no-op when the array is already `(n_samples, n_quantiles)` and correctly reshapes a 1D flat array, maintaining backward compatibility with XGBoost <3.2.

## Investigation Notes

- **`mean_pinball_loss`**: already used `np.reshape(y_pred, [-1, n_quantiles])` which handles both shapes correctly — no change needed
- **`metrics_probabilistic.py`**: no changes needed, `sample_weight_eval_set` verified working correctly with XGBoost 3.2
- **`xgboost_forecaster.py`**: no changes needed
- **`pyproject.toml`**: already at `>=3,<4` on this branch — no change needed

### CODE REVIEW

**Review by egordm — APPROVED** [review]
LGTM, both affected objective functions are correctly fixed and the pattern is consistent. Approving.

**Comment by egordm** [comment]
Hi @Soutehkeshan, thanks so much for picking this up and submitting a fix so quickly! Really appreciated.

The code change looks correct. Both `arctan_loss_multi_objective` and `pinball_loss_multi_objective` are now handled, and the existing tests pass. The only thing missing is test coverage for the new 2D input shape so we have a regression guard going forward. To avoid blocking your PR, I've added those tests in a separate follow-up (PR #867) that we'll merge right after this one.

On the DCO check failing: if you haven't already, you can either add a sign-off by amending your commit (`git commit --amend --signoff`) or we can handle it with a squash-merge on our side. No worries either way, we'll sort it out.

Looking forward to seeing more contributions from you!

**Comment by Soutehkeshan** [comment]
Hi @egordm,

Thanks for your feedback. I'm not sure if sign-off is still needed considering that the PR has already been merged. If so, could you please handle it with a squash-merge on your side?

Looking forward to contributing more.
